### PR TITLE
WIP Adds left and right widgets area

### PR DESCRIPTION
**Pull request by jweill-aws**
Related issue: #272 

This change adds left and right widgets areas to RetroLab and enables the `toc` extension which should be available in the left area.

It is WIP for a few reasons:
1. There is a show/hide menu item for the left and right areas, but it does not yet include a cascading submenu for each widget in these areas.
2. ~The primary tree view in RetroLab is not presently centered as it was before this change.~ (fixed)
3. ~The `toc-extension` is not on the same 3.2.x branch as the other extensions are (blocked on https://github.com/jupyterlab/jupyterlab/pull/11421)~ (fixed)

### CODE REVIEW

**Comment by jtpio** [comment]
Thanks @jweill-aws for starting this!

File: app/webpack.config.js [hunk 1]
```diff
@@ -175,6 +175,9 @@ module.exports = [
         name: 'CORE_FEDERATION',
         shared: createShared(data)
       })
-    ]
+    ],
+    resolve: {
+      fallback: { util: require.resolve('util/') }
```
Comment (jtpio) [inline app/webpack.config.js:180]: Wondering why this fallback is needed. Looks like this is not needed in JupyterLab? (after quickly checking the [webpack.config.js](https://github.com/jupyterlab/jupyterlab/blob/d860b89bff1ec84309314cc686adeb1b72fc9d09/builder/src/webpack.config.base.ts#L104-L110))

File: packages/application/package.json [hunk 2]
```diff
@@ -44,7 +44,7 @@
   },
   "dependencies": {
     "@jupyterlab/application": "^3.2.0",
-    "@jupyterlab/coreutils": "^5.2.0",
+    "@jupyterlab/coreutils": "^3.2.0",
```
Comment (jtpio) [inline packages/application/package.json:47]: Looks like an unintentional change? I think it should stay on `^5.2.0`:

```suggestion
    "@jupyterlab/coreutils": "^5.2.0",
```

**Comment by jtpio** [comment]
JupyterLab 3.2.3 was released yesterday: https://github.com/jupyterlab/jupyterlab/releases/tag/v3.2.3

So we can update this PR to the new packages with:

```
jlpm run update:dependency --regex @jupyterlab/ latest
```

File: app/webpack.config.js (hunk 1 above)
Comment (jweill-aws) [reply app/webpack.config.js:180]: After removing this and rebuilding, I didn't reproduce the error above. Reverting this change.

File: app/package.json [hunk 3]
```diff
@@ -57,6 +57,8 @@
     "@jupyterlab/terminal-extension": "~3.2.0",
     "@jupyterlab/theme-dark-extension": "~3.2.0",
     "@jupyterlab/theme-light-extension": "~3.2.0",
+    "@jupyterlab/toc": "~5.3.0",
+    "@jupyterlab/toc-extension": "~5.3.0-alpha.15",
```
Comment (jtpio) [inline app/package.json:61]: We can now use the latest stable versions:

```suggestion
    "@jupyterlab/toc": "~5.2.3",
    "@jupyterlab/toc-extension": "~5.2.3",
```

Comment (jtpio) [reply app/package.json:61]: Also these resolutions can be automatically generated with the `jlpm run integrity` script (after removing the previous `node_modules` to clean up)

**Comment by jweill-aws** [comment]
`toc-extension` requires a layout restorer, but RetroLab does not provide one. I have a PR out (https://github.com/jupyterlab/jupyterlab/pull/11445) to address an issue (https://github.com/jupyterlab/jupyterlab/pull/11444) to make this extension compatible with RetroLab.

**Comment by jweill-aws** [comment]
This is blocked at the moment because the changes I made in https://github.com/jupyterlab/jupyterlab/pull/11445 and the changes @fcollonval made in https://github.com/jupyterlab/jupyterlab/pull/11461 are not yet available in NPM. I tried using a `link` pseudo-version schema in `app/package.json` to incorporate the `toc` and `toc-extension` packages from my local workspace of `jupyterlab`, but `webpack` gives me errors such as the following when I try to build with the local override in place.

```
ERROR in resolving fallback for shared module @jupyterlab/toc-extension
Module not found: Error: Can't resolve '@jupyterlab/toc-extension' in '/Users/jweill/git/retrolab/app/build'
resolve '@jupyterlab/toc-extension' in '/Users/jweill/git/retrolab/app/build'
  Parsed request is a module
  using description file: /Users/jweill/git/retrolab/app/package.json (relative path: ./build)
    Field 'browser' doesn't contain a valid alias configuration
    resolve as module
      /Users/jweill/git/retrolab/app/build/node_modules doesn't exist or is not a directory
      looking for modules in /Users/jweill/git/retrolab/app/node_modules
        single file module
          using description file: /Users/jweill/git/retrolab/app/package.json (relative path: ./node_modules/@jupyterlab/toc-extension)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/app/node_modules/@jupyterlab/toc-extension doesn't exist
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/app/node_modules/@jupyterlab/toc-extension.js doesn't exist
            .json
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/app/node_modules/@jupyterlab/toc-extension.json doesn't exist
            .wasm
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/app/node_modules/@jupyterlab/toc-extension.wasm doesn't exist
        /Users/jweill/git/retrolab/app/node_modules/@jupyterlab/toc-extension doesn't exist
      looking for modules in /Users/jweill/git/retrolab/node_modules
        single file module
          using description file: /Users/jweill/git/retrolab/package.json (relative path: ./node_modules/@jupyterlab/toc-extension)
            no extension
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension is not a file
            .js
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension.js doesn't exist
            .json
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension.json doesn't exist
            .wasm
              Field 'browser' doesn't contain a valid alias configuration
              /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension.wasm doesn't exist
        existing directory /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension
          using description file: /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/package.json (relative path: .)
            using description file: /Users/jweill/git/retrolab/package.json (relative path: ./node_modules/@jupyterlab/toc-extension)
              no extension
                Field 'browser' doesn't contain a valid alias configuration
                /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension is not a file
              .js
                Field 'browser' doesn't contain a valid alias configuration
                /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension.js doesn't exist
              .json
                Field 'browser' doesn't contain a valid alias configuration
                /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension.json doesn't exist
              .wasm
                Field 'browser' doesn't contain a valid alias configuration
                /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension.wasm doesn't exist
              as directory
                existing directory /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension
                  using description file: /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/package.json (relative path: .)
                    use ./lib/index.js from main in package.json
                      using description file: /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/package.json (relative path: ./lib/index.js)
                        no extension
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/lib/index.js doesn't exist
                        .js
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/lib/index.js.js doesn't exist
                        .json
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/lib/index.js.json doesn't exist
                        .wasm
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/lib/index.js.wasm doesn't exist
                        as directory
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/lib/index.js doesn't exist
                    using path: /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/index
                      using description file: /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/package.json (relative path: ./index)
                        no extension
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/index doesn't exist
                        .js
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/index.js doesn't exist
                        .json
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/index.json doesn't exist
                        .wasm
                          Field 'browser' doesn't contain a valid alias configuration
                          /Users/jweill/git/retrolab/node_modules/@jupyterlab/toc-extension/index.wasm doesn't exist
      /Users/jweill/git/node_modules doesn't exist or is not a directory
      /Users/jweill/node_modules doesn't exist or is not a directory
      /Users/node_modules doesn't exist or is not a directory
      /node_modules doesn't exist or is not a directory

webpack 5.37.1 compiled with 1 error in 10639 ms
```

**Comment by fcollonval** [comment]
I think the easiest way forward is to publish a new version of JLab 3.2.x

**Comment by jtpio** [comment]
> So we can update this PR to the new packages with:
> 
> ```
> jlpm run update:dependency --regex @jupyterlab/ latest
> ```

@jweill-aws rerunning this command should pick up the latest release.

**Comment by jtpio** [comment]
Looks like it now builds and loads on Binder :+1: 

image

**Comment by jtpio** [comment]
Probably we now need to define our own plugin that would depend on `ITableOfContentsRegistry`, and use `app.shell.currentWidget` as the widget (the current widget does not change at the moment in retrolab).

Since the upstream logic does not apply in RetroLab: https://github.com/jupyterlab/jupyterlab/blob/7aba384baa09393e98432619b3355f57d7529012/packages/toc-extension/src/index.ts#L199-L202

**Comment by fcollonval** [comment]
> Probably we now need to define our own plugin that would depend on `ITableOfContentsRegistry`, and use `app.shell.currentWidget` as the widget (the current widget does not change at the moment in retrolab).

In fact that will not be as simple because unfortunately we don't have access to the `TableOfContents` object that needs to be set :disappointed: 

https://github.com/jupyterlab/jupyterlab/blob/7aba384baa09393e98432619b3355f57d7529012/packages/toc-extension/src/index.ts#L225

**Comment by jweill-aws** [comment]
Right now, the Command Palette is added, then the menu item is added, then the `toc` widget is added, but the menu does not update. Selecting the Command Palette's menu item causes the `toc` widget to appear.

**Comment by hbcarlos** [comment]
Hi everyone! And thank you soo much for the effort put on this, is looking incredible!
I have been following this PR since I want to use side panels on retrolab for an extension (voila-gridstack).

I would like to give some UX feedback on the side panels.
I think by default they should be hidden and only show up when the user triggers an action that requires something on the side panel. Also, regarding the design, we could divide the shell into three parts to keep the notebook always in the middle and then create the box panel for the side panels, the shell would be divided like in the following image:

notebook

This way, we can constrain the width of the main area widget and topbar together while adding panels on the side.
Regarding the left and right areas of the topbar, I think it's fine to remove that since it is always empty.

**Comment by jweill-aws** [comment]
@hbcarlos Thank you for your feedback! The concept I'm working with right now (still a work in progress) is to hide the sidebars on initial load, but make the Table of Contents widget available from the View menu. When it's exposed, the TOC widget would show up on the left side, pushing the main area to the right, narrowing it if necessary, but not overlapping. Right-area widgets would work in the same way, and a left and right area widget could be visible simultaneously. The screen shots below show a WIP concept of a workbook in RetroLab with no widgets open and with a left-panel widget open. The side panels are also resizable.

image

image

**Comment by jweill-aws** [comment]
Current open issues for me:

❓How can I make the menu plugin listen for changes to the widget content of the left/right area?
❓Why is the TOC extension not updating when I select something in the document or when I modify the document?

**Comment by jtpio** [comment]
> Current open issues for me:

I think this is because of the logic mentioned by @fcollonval above:

https://github.com/jupyterlab/jupyterlab/blob/7aba384baa09393e98432619b3355f57d7529012/packages/toc-extension/src/index.ts#L225

So we might need to rework the upstream plugin a bit more. I'll draft something and open a PR upstream.

In the meantime we can temporarily copy some the toc logic here in a temp plugin, so we are not blocked by a release of the upstream packages.

**Comment by jtpio** [comment]
@jweill-aws I investigated this a bit in https://github.com/jupyterlab/jupyterlab/pull/11530.

As @fcollonval mentioned above, the widget and registry are quite coupled in the upstream plugin. This likely requires bigger changes upstream (maybe backward incompatible) so we can reuse the plugin as is.

In the meantime one way forward would be to define our own `@retrolab/notebook-extension:toc` plugin in https://github.com/jupyterlab/retrolab/blob/main/packages/notebook-extension/src/index.ts

It would be similar to the upstream plugin: https://github.com/jupyterlab/jupyterlab/blob/0472d1ad59a69f494660654dca382dbd7a810364/packages/toc-extension/src/index.ts#L234-L248 (some duplicated code for now).
Although we could only depend on the `INotebookTracker` since we plan to show the table of contents for notebooks only for now.

**Comment by jweill-aws** [comment]
@jtpio @fcollonval I'm back in the (home) office today after taking last week off for the U.S. Thanksgiving holiday. I've been reviewing the added comments on this and other issues. If we define our own `@retrolab/notebook-extension:toc` plugin in the `notebook-extension` package, what aspects of the existing TOC plugin would I need to recreate, and with what modifications? This week is somewhat busy for me with AWS re:invent keynotes most mornings, but I'm curious about what the next steps are for this issue.

**Comment by fcollonval** [comment]
Hey Jason, it should be as simple as duplicating [that file](https://github.com/jupyterlab/jupyterlab/blob/0472d1ad59a69f494660654dca382dbd7a810364/packages/toc-extension/src/index.ts) mentioned by Jeremy and remove in the activate function what is not relevant for retrolab + setting the current widget as the notebook like in https://github.com/jupyterlab/jupyterlab/blob/0472d1ad59a69f494660654dca382dbd7a810364/packages/toc-extension/src/index.ts#L225

**Comment by ellisonbg** [comment]
@fcollonval I think the goal of this effort is broader than getting the TOC extension to work in retrolab. The goal is to get any extension that uses the L/R areas to work in retrolab, without modification, copying code, etc. Thus, we can duplicate files, etc. that will create long term technical debt and doesn't help us get all the other extensions that use the L/R panels (debugger, properties, etc.) into retrolab. I will work with @jweill-aws to come up with a broader plan.

**Comment by jweill-aws** [comment]
I should be able to modify the `toc-extension` to accept an `IShell` object, but despite `application/src/frontend.ts` exporting the `IShell` interface, the compiled `frontend.d.ts` file doesn't declare this interface as exported.

The only additional method I need is `currentChanged`, which is not part of the `IShell` interface, but which is implemented (with different signatures) in both `RetroShell` and `LabShell`:

RetroShell:

```
  get currentChanged(): ISignal<RetroShell, void>;
```

LabShell:

```
  export type IChangedArgs = FocusTracker.IChangedArgs<Widget>;
  ...
  get currentChanged(): ISignal<this, ILabShell.IChangedArgs>;
```

It seems like if we can declare `currentChanged` in a consistent way, then move this definition up to `IShell`, and ensure that this interface is exported properly, we can use it in the `toc` extension.

**Comment by jtpio** [comment]
For sure extensions should just work the same in both lab and retro. In fact most of them already do.

I think the point here was to be able to make some progress on that particular TOC extension, and lay the foundations for the left and right areas so other extensions can also be used. So we can merge this PR and iterate.

TOC is still a bit special for now because of the way it currently works. See https://github.com/jupyterlab/jupyterlab/pull/11530 for more info. Making it compatible with RetroLab probably requires some breaking changes, and since it is part of the `jupyterlab` repo this adds significant delays before being able to consume the changes it here in RetroLab.

Here is an example of `jupyterlab-github` as another third-party extension added to the `left` area in RetroLab, using a local build of this PR locally and a minor change upstream to make the extension compatible (https://github.com/jupyterlab/jupyterlab-github/pull/128):

https://user-images.githubusercontent.com/591645/144586705-0e8b3b9f-b001-49c6-969d-253b7209f881.mp4

**Comment by jweill-aws** [comment]
I'm still seeing the command palette appear in the context menu as an extension, but not the TOC extension. When I choose this option, though, the left pane appears with the TOC extension in it.

image

**Comment by jtpio** [comment]
> I'm still seeing the command palette appear in the context menu as an extension

Probably this is because the palette is added to the left area here: 

https://github.com/jupyterlab/jupyterlab/blob/5829e22634c8ee7704e805b7bbb5cb15ca7ca8b6/packages/apputils-extension/src/palette.ts#L103

Which seems to be the case even though by default the palette is displayed as a modal now (before the default was to add it to the left area).

**Comment by jweill-aws** [comment]
@jtpio I noticed that and was confused. Is it safe to remove the palette from the left area? Does it ever display there in JupyterLab?

**Comment by bollwyvl** [comment]
Yes, the palette can be shown by config and restores useful metadata
(category). It's actually more accessible, as the links are less sensitive
to focus/mouse movements.

**Comment by jweill-aws** [comment]
Thanks @bollwyvl — didn't realize that there was a config option to make the palette visible. This raises two questions:
1. Is this option enabled by default in RetroLab? If not, why is the menu item appearing?
2. Why does the contextual menu only show one option for the left sidebar? Why isn't the TOC extension also appearing?

**Comment by jweill-aws** [comment]
@jtpio I noticed when I was looking into the menu item issue that the menu gets generated once, after the command palette plugin has been activated, but it doesn't get regenerated with two items after the TOC plugin has been activated. I'm concerned that the lack of a notebook tracker or a widget tracker in RetroLab's left pane may be related to this: there is no `widgetAdded` event that the menu plugin can listen to. Should we be using a widget tracker with the left sidebar? Is there another pattern that would result in the menu getting updated?

**Comment by jtpio** [comment]
One way to handle this for now could be to wait for `app.restored` before creating the menu entries.

> I'm concerned that the lack of a notebook tracker or a widget tracker in RetroLab's left pane

RetroLab supports notebook trackers and other widget trackers. For example the notebook tracker plugin is added here:

https://github.com/jupyterlab/retrolab/blob/6b47f0cf62571c5c53db3b71f1d6ae4acb0f072a/app/index.js#L138

Although it's true there is no way to know a widget has been added to the left / right areas. But isn't that the case for the `LabShell` too?

**Comment by jweill-aws** [comment]
> One way to handle this for now could be to wait for `app.restored` before creating the menu entries.

@jtpio Thanks! I've updated this PR in this way. Still not seeing the toc extension load, but at least it appears as its own item in the menu now. Progress!

**Comment by jweill-aws** [comment]
At the February 2 JupyterLab meeting, someone mentioned using a "live inspector" extension in lieu of the `toc-extension` to try out the left and right areas. Which should I use? In JupyterLab, I found `property-inspector` which requires `ILabShell` (not compatible with the `RetroShell` used by RetroLab) and `inspector` which doesn't seem to be added to any shell area.

**Comment by jtpio** [comment]
> I found `property-inspector` which requires `ILabShell` (not compatible with the `RetroShell` used by RetroLab)

Right for reference `IPropertyInspectorProvider` is provided here and the plugin does depends on `ILabShell`:

https://github.com/jupyterlab/jupyterlab/blob/2c2d83fc4b39f292ed12dc9d90c32a2cdb606943/packages/application-extension/src/index.tsx#L994-L1032

Maybe it can be made optional so `ILabShell` is only used for listening to the `currentChanged` signal.

**Comment by bollwyvl** [comment]
So many inspectors!

I was actually referring to the thing [called `@jupyterlab/inspector:IInspector`](https://github.com/jupyterlab/jupyterlab/blob/master/packages/inspector-extension/src/index.ts) that now appears as a main area widget under _Contextual Help_.

This is an implementation of a top-level [jupyter message](https://jupyter-client.readthedocs.io/en/stable/messaging.html#introspection) and is a great interactive computing "passive" learning tool, as it doesn't require any keystrokes, etc. to see a stream of new information

This was originally a (right) sidebar, and was moved to the main area to be more flexible.

In nb6, it appears in the "pager" area at the bottom of the page... i personally never liked that position, and would not be sad at all if it appeared next to the code-of-interest, and consumed some of the whitespace.

It also looks to only require `ITranslator` which is good for business.

**Comment by jtpio** [comment]
@jweill-aws FYI the RetroLab code base has now been integrated in the Notebook repo: https://github.com/jupyter/notebook/

Would you like to resume continue this PR over there?

Thanks!

**Comment by jweill-aws** [comment]
Closing in favor of https://github.com/jupyter/notebook/pull/6327.

**Pull request closed without merging**